hail/src/main/scala/is/hail/backend/spark/LowerTableIR.scala

```diff
@@ -1,29 +1,209 @@
 package is.hail.backend.spark
 
-import is.hail.annotations.RegionValue
-import is.hail.backend.Binding
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream}
+
+import is.hail.annotations._
+import is.hail.cxx
 import is.hail.expr.ir._
+import is.hail.expr.types.physical.PTuple
 import is.hail.expr.types.virtual._
+import is.hail.io.CodecSpec
+import is.hail.nativecode._
 import is.hail.rvd.{RVDPartitioner, RVDType}
 import is.hail.utils._
+import org.apache.spark.SparkContext
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.Row
 
-case class SparkCollect(
-  child: SparkStage)
+object SparkPipeline {
+  def local(body: IR): SparkPipeline = SparkPipeline(Map.empty[String, SparkStage], body)
+}
+
+case class SparkPipeline(stages: Map[String, SparkStage], body: IR) {
+  def codecSpec: CodecSpec = CodecSpec.defaultUncompressed
+
+  def typ: Type = body.typ
+
+  def execute(sc: SparkContext, region: Region): (PTuple, Long) = {
+    val fields: IndexedSeq[(String, Type)] =
+      stages.map { case (name, stage) => (name, TArray(stage.body.typ)) }.toFastIndexedSeq
+
+    val inputType = TStruct(fields: _*)
+    val rvb = new RegionValueBuilder(region)
+    rvb.start(inputType.physicalType)
+    rvb.startStruct()
+    stages.foreach { case (_, stage) => stage.collect(sc, rvb) }
+    rvb.endStruct()
 
-case class SparkShuffle(
-  child: SparkStage)
+    val ref = Ref(genUID(), inputType)
+    val node = Subst(body, Env[IR](fields.map { case (name, _) => name -> GetField(ref, name) }.toFastSeq : _*))
+    val (typ, f) = Compile[Long, Long](ref.name, inputType.physicalType, MakeTuple(FastSeq(node)))
+    (typ.asInstanceOf[PTuple], f(0)(region, rvb.end(), false))
+  }
+}
+
+case class SparkShuffle(child: SparkStage)
+
+case class SparkBinding(name: String, value: SparkPipeline)
 
 case class SparkStage(
-  globals: List[Binding],
+  globals: List[SparkBinding],
   rvdType: RVDType,
   partitioner: RVDPartitioner,
   rdds: Map[String, RDD[RegionValue]],
   contextType: Type, context: Array[Any],
-  body: IR)
+  body: IR) {
+
+  def codecSpec: CodecSpec = CodecSpec.defaultUncompressed
+
+  def collect(sc: SparkContext, rvb: RegionValueBuilder): Unit = {
+    assert(rdds.isEmpty)
+
+    val region = rvb.region
+
+    val gType = TStruct(globals.map { b => b.name -> b.value.typ }: _*)
+    val rvb2 = new RegionValueBuilder(region)
+    rvb2.start(gType.physicalType)
+    rvb2.startStruct()
+    globals.foreach { case SparkBinding(_, value) =>
+      val (typ, off) = value.execute(sc, region)
+      if (typ.isFieldDefined(region, off, 0))
+        rvb2.addRegionValue(typ.types(0), region, typ.loadField(region, off, 0))
+      else
+        rvb2.setMissing()
+    }
+    rvb2.endStruct()
+
+    val baos = new ByteArrayOutputStream()
+    val gEncoder = codecSpec.buildEncoder(gType.physicalType)(baos)
+    gEncoder.writeRegionValue(region, rvb2.end())
+    gEncoder.flush()
+    val globsBC = baos.toByteArray
```

Comment:
I think this can be replaced by `RegionValue.toBytes`, and analogously `fromBytes` for the decoder.